fvm/evm/offchain/sync/replay.go

```diff
@@ -0,0 +1,250 @@
+		return err
+	}
+
+	var res *types.Result
+	// check if the transaction payload is actually from a direct call,
+	// which is a special state transition in Flow EVM.
+	if txEvent.TransactionType == types.DirectCallTxType {
+		call, err := types.DirectCallFromEncoded(txEvent.Payload)
+		if err != nil {
+			return fmt.Errorf("failed to RLP-decode direct call [%x]: %w", txEvent.Payload, err)
+		}
+		res, err = bv.DirectCall(call)
+		if err != nil {
+			return fmt.Errorf("failed to execute direct call [%x]: %w", txEvent.Hash, err)
+		}
+	} else {
+		gethTx := &gethTypes.Transaction{}
+		if err := gethTx.UnmarshalBinary(txEvent.Payload); err != nil {
+			return fmt.Errorf("failed to RLP-decode transaction [%x]: %w", txEvent.Payload, err)
+		}
+		res, err = bv.RunTransaction(gethTx)
+		if err != nil {
+			return fmt.Errorf("failed to run transaction [%x]: %w", txEvent.Hash, err)
+		}
+	}
+
+	// validate results
+	if validate {
+		if err := ValidateResult(res, txEvent); err != nil {
+			return fmt.Errorf("transaction replay failed (txHash %x): %w", txEvent.Hash, err)
+		}
+	}
+
+	return nil
+}
+
+func ValidateResult(
+	res *types.Result,
+	txEvent *events.TransactionEventPayload,
+) error {
+
+	// we should never produce invalid transaction, since if the transaction was emitted from the evm core
+	// it must have either been successful or failed, invalid transactions are not emitted
+	if res.Invalid() {
+		return fmt.Errorf("invalid transaction: %w", res.ValidationError)
+	}
+
+	// check gas consumed
+	if res.GasConsumed != txEvent.GasConsumed {
+		return fmt.Errorf("gas consumption mismatch %d != %d", res.GasConsumed, txEvent.GasConsumed)
+	}
+
+	// check error msg
+	if errMsg := res.ErrorMsg(); errMsg != txEvent.ErrorMessage {
+		return fmt.Errorf("error msg mismatch %s != %s", errMsg, txEvent.ErrorMessage)
+	}
+
+	// check encoded logs
+	encodedLogs, err := res.RLPEncodedLogs()
+	if err != nil {
+		return fmt.Errorf("failed to RLP-encode logs: %w", err)
+	}
+	if !bytes.Equal(encodedLogs, txEvent.Logs) {
+		return fmt.Errorf("encoded logs mismatch %s != %s", encodedLogs, txEvent.ErrorMessage)
+	}
+
+	// check deployed address
+	if deployedAddress := res.DeployedContractAddressString(); deployedAddress != txEvent.ContractAddress {
+		return fmt.Errorf("deployed address mismatch %s != %s", deployedAddress, txEvent.ContractAddress)
+	}
+
+	// check the state change checksum
+	// if empty checksum skip (supporting blocks before checksum integration)
+	if checksum := res.StateChangeChecksum(); txEvent.StateUpdateChecksum != emptyChecksum &&
+		checksum != txEvent.StateUpdateChecksum {
+		return fmt.Errorf("state change checksum mismatch %x != %x", checksum, txEvent.StateUpdateChecksum)
+	}
+
+	return nil
+}
+
+// CreateBlockContext creates a block context using the passed blocks
+func CreateBlockContext(
+	chainID flow.ChainID,
+	blocks *blocks.Blocks,
+	tracer *gethTracer.Tracer,
+) (types.BlockContext, error) {
+	bm, err := blocks.LatestBlock()
```

Comment:
Why do we use the latest block metadata here?
If `ReplayBlockExecution` is called with `blockEvent *events.BlockEventPayload`, which is say 50 blocks behind the latest block, shouldn't we replay all transactions of `blockEvent.Height` at the known state of `blockEvent.Height - 1`?

Reply:
I moved it to the logic and renamed it to be latestBlockContext, yes we need to replay all transactions of the previous heights. but this still works given the latest block here is updated per block.

Reply:
Note that the latestBlock is relative to the time we execute things, the latest block meta at block 10 returns info on the block 10 even the chain is at block 50.

Reply:
Ah, makes sense now :+1: